invenio_preservation_sync/api.py

```diff
@@ -0,0 +1,141 @@
+            harvest_timestamp=harvest_timestamp,
+            archive_timestamp=archive_timestamp,
+            uri=uri,
+            path=path,
+            event_id=event_id,
+            description=description,
+        )
+
+        db.session.add(preservation)
+        db.session.commit()
+        return preservation
+
+    @classmethod
+    def get_by_record_id(cls, record_id=None, latest=False):
+        """Get Preservation Info by record id."""
+        preservation = PreservationInfo.query.filter_by(record_id=record_id).order_by(
+            PreservationInfo.revision_id.desc(),
+            PreservationInfo.archive_timestamp.desc(),
+            PreservationInfo.created.desc(),
+        )
+        if latest:
+            return preservation.first()
+        return preservation.all()
+
+    @classmethod
+    def get_existing_preservation(
+        cls,
+        record_id=None,
+        revision_id=None,
+        archive_timestamp=None,
+    ):
+        """Return Preservation Info if it already exists."""
+        preservation = PreservationInfo.query.filter_by(
+            record_id=record_id,
+            revision_id=revision_id,
+            archive_timestamp=archive_timestamp,
+        ).first()
+        return preservation
+
+    @classmethod
+    def update_existing_preservation(
+        cls,
+        preservation,
+        status=None,
+        harvest_timestamp=None,
+        uri=None,
+        path=None,
+        event_id=None,
+        description=None,
+    ):
+        """Update existing preservation."""
+        status = PreservationInfoAPI._convert_status(status)
+
+        if (
+            preservation.status == status
+            and preservation.harvest_timestamp == parse(harvest_timestamp)
+            and preservation.uri == uri
+            and preservation.path == path
+            and preservation.description == description
+        ):
+            raise PreservationAlreadyReceivedError(preservation)
+
+        if status:
+            preservation.status = status
+
+        if harvest_timestamp:
+            preservation.harvest_timestamp = harvest_timestamp
+
+        if uri:
+            preservation.uri = uri
+
+        if path:
+            preservation.path = path
+
+        if description:
+            preservation.description = description
+
+        preservation.event_id = event_id
+
+        db.session.commit()
+
+    @classmethod
+    def _convert_status(cls, value):
+        """Convert the status of the preservation info.
+
+        Valid examples: "P", "Preserved", "preserved"
+        """
+        if isinstance(value, PreservationStatus):
+            return value
+        elif isinstance(value, str):
+            upper_value = value.upper()
+            if PreservationStatus.has_key(upper_value):
+                return PreservationStatus[upper_value]
+            return PreservationStatus(value.upper())
+        else:
+            raise TypeError("Value must be a PreservationStatus or a string.")
```

Comment:
You might want to have a more specific error here, or wrap the `TypeError` in the resource layer: if I am sending something wrong in my request, and I get this error, the msg is not really helping me to understand what I have done wrong.